src/builder.cpp

```diff
@@ -0,0 +1,312 @@
+#include <builder.hpp>
+#include <charconv>
+#include <exception.hpp>
+#include <log.hpp>
+#include <parser/common.hpp>
+#include <parser/parser.hpp>
+#include <string_view>
+
+namespace ddwaf {
+
+constexpr builder::change_state operator|(builder::change_state lhs, builder::change_state rhs)
+{
+    return static_cast<builder::change_state>(
+        static_cast<std::underlying_type<builder::change_state>::type>(lhs) |
+        static_cast<std::underlying_type<builder::change_state>::type>(rhs));
+}
+
+constexpr builder::change_state operator&(builder::change_state lhs, builder::change_state rhs)
+{
+    return static_cast<builder::change_state>(
+        static_cast<std::underlying_type<builder::change_state>::type>(lhs) &
+        static_cast<std::underlying_type<builder::change_state>::type>(rhs));
+}
+
+std::shared_ptr<ruleset> builder::build(parameter object, ruleset_info &info)
+{
+    parameter::map ruleset = object;
+
+    unsigned version = parser::parse_schema_version(ruleset);
+    if (version == 1) {
+        ddwaf::ruleset rs;
+        parser::v1::parse(ruleset, info, rs, limits_);
+        return std::make_shared<ddwaf::ruleset>(std::move(rs));
+    }
+
+    if (version != 2) {
+        DDWAF_ERROR("incompatible ruleset version %u.x", version);
+        throw unsupported_version();
+    }
+
+    return build_helper(ruleset, info);
+}
+
+namespace {
+
+std::set<rule::ptr> target_to_rules(const std::vector<parser::rule_target_spec> &targets,
+    const std::unordered_map<std::string_view, rule::ptr> &rules, const rule_tag_map &rules_by_tags)
+{
+    std::set<rule::ptr> rule_targets;
+    if (!targets.empty()) {
+        for (const auto &target : targets) {
+            if (target.type == parser::target_type::id) {
+                auto rule_it = rules.find(target.rule_id);
+                if (rule_it == rules.end()) {
+                    continue;
+                }
+                rule_targets.emplace(rule_it->second);
+            } else if (target.type == parser::target_type::tags) {
+                auto current_targets = rules_by_tags.multifind(target.tags);
+                rule_targets.merge(current_targets);
+            }
+        }
+    } else {
+        // An empty rules target applies to all rules
+        for (const auto &[id, rule] : rules) { rule_targets.emplace(rule); }
+    }
+    return rule_targets;
+}
+
+} // namespace
+
+std::shared_ptr<ruleset> builder::build_helper(parameter::map &root, ruleset_info &info)
+{
+    // Load new rules, overrides and exclusions
+    auto state = load(root, info);
+
+    if (state == change_state::none) {
+        return {};
+    }
+
+    constexpr change_state rule_update =
+        change_state::rules | change_state::data | change_state::overrides;
+    constexpr change_state filters_update = rule_update | change_state::filters;
+    constexpr change_state manifest_update = change_state::rules | change_state::filters;
+
+    if ((state & rule_update) != change_state::none) {
```

Comment:
The fact that merely changing data (a relatively frequent occurrence) causes all this to be recalculated is quite problematic IMO.

Reply:
As discussed privately, I'll improve this with a dynamic rule processor map in the next iteration (1.8.1).

Reply:
I managed to implement the dynamic rule processor map so the rules don't need to be recomputed any longer after a data update.